Fix issue #30658 by checking explicitly for 'null' referrer

JackMc opened the pull request:
### Summary

Currently, when a POST request with request forgery checking enabled occurs, Rails checks the 
`request.origin` to see if it is either `nil` (not sent) or equal to the expected origin. As mentioned in 
issue #30658, Chrome has begun serving a referrer of `null` when `no-referrer` is specified in the
referrer policy header. Because of this, all authenticity-checking POST requests originating on a 
page with `no-referrer` enabled will fail. This PR adds code that checks for the explicit string "null"
in the referrer and allows it through.

JackMc committed:
Add a better error message when a "null" Origin header occurs

JackMc commented:
I'm testing this in Firefox to see if it happens there too

sgrif commented:
Since this is identical to #28304 but with tests, can you make sure the original author is given credit? I need to do more research on what's up with Chrome before merging this, I don't want to add a workaround to the framework if it's a bug in Chrome

JackMc commented:
Added @sitdot3291 as a referenced author

JackMc commented:
I did a couple experiments on the differences in behaviour between Firefox and Chrome here. I gathered the following:

* When you ask for `no-referrer` referrer policy on the originating page in Chrome, a same-origin POST will generate a `null` referrer, but a CORS request will always give the origin
* When you ask for `origin` referrer policy on the originating page in Chrome, both a same-origin POST and a cross-origin request will give the origin
* Regardless of if you ask for `no-referrer` or `origin` on Firefox, the behaviour remains the same: **no origin header is sent** when you do a same-origin POST

I've found a few conflicting sources regarding if Chrome's or Firefox's or both's behaviour is acceptable. Going to do more reading in the RFC.

sgrif commented:
This is definitely a bug in Chrome. I couldn't find an open bug report there (though there is https://bugs.chromium.org/p/chromium/issues/detail?id=635400 which is similar). Probably worth opening a bug report on Chrome if nothing else. Given that Chrome is evergreen, and all indications point to this being a bug in the browser, I'd rather not introduce a framework workaround for it. I'm inclined to close this, the issue you're referencing, and the original issue #28299, but I'll let @pixeltrix make the final call since he chimed in on the initial report.

r? @pixeltrix

sgrif commented:
So after doing further digging, this is not a bug in chrome. It could be considered a bug in Firefox, (see https://bugzilla.mozilla.org/show_bug.cgi?id=446344), but basically the reason this is happening is due to a rather vague line in [the `Origin` spec](https://tools.ietf.org/html/rfc6454):

> Whenever a user agent issues an HTTP request from a "privacy-sensitive" context, the user agent MUST send the value "null" in the Origin header field.

It's likely that most people setting the `no-referrer` value probably want `strict-same-origin` instead. If they don't, they'll need to change the value of the `forgery_protection_origin_check` to `false` and opt out of this. Rails is currently doing the right thing [as described in the original proposal of the `Origin` header for CSRF protection](http://seclab.stanford.edu/websec/csrf/csrf.pdf)

> If the Origin header is present, the server must reject any requests whose Origin header contains an undesired value (including null).

@JackMc I think we can probably provide a more helpful error message in this case. If you're willing to update this PR to raise an Exception that explains the problem, I would merge that. Otherwise, this PR can be closed, as it won't be accepted as-is.

JackMc commented:
Sorry for the long delay. I've edited the PR and tests to give a detailed error message regarding what the user has to do to make the null referrer work. I wasn't sure if this falls under InvalidAuthenticityToken (that seems to be what's thrown when the origin is wrong otherwise.

sgrif commented on actionpack/lib/action_controller/metal/request_forgery_protection.rb:
```diff
@@ -419,6 +419,12 @@ def protect_against_forgery? # :doc:
         if forgery_protection_origin_check
           # We accept blank origin headers because some user agents don't send it.
+          raise InvalidAuthenticityToken, "The browser returned a 'null' origin for a request with origin-based" \
+          " forgery protection turned on. This usually means you have the 'no-referrer' Referrer-Policy header " \
```
This is as nit-picky as it gets, but it bothers me that this is the only line with a space at the beginning rather than the end. Is it worth just doing a `strip_heredoc` here?

sgrif commented on actionpack/lib/action_controller/metal/request_forgery_protection.rb:
```diff
@@ -419,6 +419,12 @@ def protect_against_forgery? # :doc:
+          "enabled, or that you the request came from a site that refused to give its origin. This makes it " \
+          "impossible for Rails to verify the source of the requests. If you are sure that this is the "\
+          "behaviour you want, you can disable origin checking by doing "\
+          "Rails.application.config.action_controller.forgery_protection_origin_check = false" if request.origin == "null"
```
Do you think most people will think to look for a postfix conditional at the end of a 6 line expression?

sgrif commented:
I have some nitpicks about the structuring of the error. Additionally, I think that the vast majority of people encountering this are better served by changing from `no-referrer` to `strict-same-origin` or `same-origin`. I think we should mention that option before mentioning how to disable origin checking.

JackMc commented:
I changed it to mention the option of changing the referrer policy to something lighter before suggesting changing the setting, and changed it to use a HEREDOC.
